R/history.R

```diff
@@ -0,0 +1,16 @@
+#' Update URL in browser's location bar
+#'
+#' This function updates the client browser's query string in the location bar.
+#' It typically is called from an observer. Note that this will not work in
+#' Internet Explorer 9 and below.
+#'
+#' @param state State to save on the history stack.
+#' @param title Not currently in use.
+#' @param url New url to display on the window.
+#' @param session A Shiny session object.
+#' @seealso \code{\link{enableBookmarking}} for examples.
+#' @export
+pushState <- function(state, title, url,
```

Comment:
Yes, totally agree. Although I think it will be confusing for users to have two functions (`updateQueryString` and `navigateQueryString`) that do exactly the same from an user perspective (change the query string)... As I look more into it, we don't really need to use `replaceState()` for bookmarking. If the goal is just to update the query string, we can achieve the same effect by calling `pushState()` (which does change the query string just the same, but also adds a new history entry onto the stack -- fortunately this is irrelevant for bookmarkable state). So I propose we get rid of the `pushState` function in Shiny altogether and use `updateQueryString()` for both scenarios, given that we change its implementation from:

```r
addMessageHandler('updateQueryString', function(message) {
    window.history.replaceState(null, null, message.queryString);
});
```
to:
```r
addMessageHandler('updateQueryString', function(message) {
    window.history.pushState(null, null, message.queryString);
});
```

We'd also have to add examples of the new usage (i.e. keep the bookmarking example copied below, but also add at least one example of how to use this for pseudo-navigation).

To make my point that it is indifferent to use `replaceState` or `pushState` when dealing with bookmarking, try the app below and toggle which line is used (`updateQueryString(url)` or `pushState(NULL, NULL, url)`) -- you'll see it works exactly the same for bookmarking purposes:

```r
ui <- function(req) {
  fluidPage(
    textInput("txt", "Text"),
    checkboxInput("chk", "Checkbox")
  )
}
server <- function(input, output, session) {
  observe({
    reactiveValuesToList(input)
    session$doBookmark()
  })
  onBookmarked(function(url) {
    # ========= TOGGLE COMMENTED OUT LINES ========== #
    # pushState(NULL, NULL, url)
    # updateQueryString(url)
  })
}
enableBookmarking("url")
shinyApp(ui, server)
```

The only problem with this is that you wouldn't be able to use `updateQueryString()` for both purposes simultaneously (which would be a pretty complicated mess even if we let the two separate functions be separate: `replaceState` for `updateQueryString` and `pushState` for `navigateQueryString`). Note that this doesn't mean you can't do bookmarking and navigation simultaneously -- it's only the very specific type of "live" bookmarking that you can't do at the same time as navigation (which, I think, makes sense that is not simultaneously supported since when you "navigate", you're presumably walking away from inputs on that "page" anyway). The only reason that there would be a good scenario to support both features simultaneously is if people wanted to have multiple-page, navigable apps and also have all inputs across all subpages to be bookmarked in the same url. Even then, I think any practical implementation of this would look like a mess and what would really preferable in those situations would be to use the "Save to server" option for bookmarking and leave the query string to handle only navigation (except, of course, for the `_state_id_` param). This is already possible with this implementation given that either:

1. it's trivial to get the correct subpage from the inputs alone (like the example app in the description of this PR);

2. the code to get the correct subpage in included in the `onRestore()` function.

So, I don't think we're really losing any substantial functionality  by doing this...

In any case, if this use case is frequent enough, I still think we should not address it here and now, because the only real way to address it would be to support proper navigation (i.e. have `app/subpage` rather than `app/?param=value`).

*Aside:* This whole exposition of the problem made me realize that when we **do** implement proper navigation, we should use modules under the hood. The namespace would be the relative url of each subpage. This would allow subpages to use the same ids for inputs and outputs and still preserve the uniqueness of ids for the final app. Even though it looks like a minor advantage, I think it could make app author's lives easier by literally using the same template for similar subpages (in the case of my example app, this would be each employee's profile).

@wch, @jcheng5, what do you think?